src/openms/source/ANALYSIS/OPENSWATH/SwathMapMassCorrection.cpp

```diff
@@ -42,19 +42,295 @@
+
+    if (im_extraction_win < 0)
+    {
+      return;
+    }
+
+    if (im_correction_function_ == "none")
+    {
+      return;
+    }
+    // if it is not none, then it must be linear
+
+    std::ofstream os_im;
+    if (!debug_im_file_.empty())
+    {
+      std::cout.precision(16);
+      os_im.open(debug_im_file_);
+      os_im << "mz" << "\t" << "im" << "\t" << "theo_im" << "\t" << "RT" << "\t" << "intensity" << std::endl;
+      os_im.precision(writtenDigits(double()));
+    }
+
+    std::vector<String> trgr_ids;
+    std::map<std::string, double> pep_im_map;
+    for (const auto& trgroup_it : transition_group_map)
+    {
+      trgr_ids.push_back(trgroup_it.first);
+    }
+    for (const auto& cmp : targeted_exp.getCompounds())
+    {
+      pep_im_map[cmp.id] = cmp.drift_time;
+    }
+
+    TransformationDescription::DataPoints data_im;
+    std::vector<double> exp_im;
+    std::vector<double> theo_im;
+#ifdef _OPENMP
+#pragma omp parallel for 
+#endif
+    for (SignedSize k = 0; k < (SignedSize)trgr_ids.size(); k++)
+    {
+      // we need at least one feature to find the best one
+      auto transition_group = transition_group_map.at(trgr_ids[k]);
+      if (transition_group->getFeatures().size() == 0)
+      {
+        continue;
+      }
+
+      // Find the feature with the highest score
+      double bestRT;
+      findBestFeature(*transition_group, bestRT);
+      // Get the corresponding SWATH map(s), for SONAR there will be more than one map
+      std::vector<OpenSwath::SwathMap> used_maps = findSwathMaps(*transition_group, swath_maps);
+
+      std::vector<OpenSwath::SwathMap> ms1_maps;
+      for (const auto& m : swath_maps) {if (m.ms1) ms1_maps.push_back(m);}
+
+      if (used_maps.empty())
+      {
+        continue;
+      }
+
+      // Get the spectrum for this RT and extract raw data points for all the
+      // calibrating transitions (fragment m/z values) from the spectrum
+      // Note that we are not using light clones of the underlying data here,
+      // so access to the data needs to be in a critical section.
+      OpenSwath::SpectrumPtr sp;
+      OpenSwath::SpectrumPtr sp_ms1;
+#ifdef _OPENMP
+#pragma omp critical
+#endif
+      {
+        if (ms1_im_)
+        {
+          sp_ms1 = OpenSwathScoring().fetchSpectrumSwath(ms1_maps, bestRT, 1, 0, 0);
+        }
+        else
+        {
+          sp = OpenSwathScoring().fetchSpectrumSwath(used_maps, bestRT, 1, 0, 0);
+        }
+      }
+
+      for (const auto& tr : transition_group->getTransitions())
+      {
+        if (ms1_im_) {continue;}
+        double intensity(0), im(0), left(tr.product_mz), right(tr.product_mz);
+
+        auto pepref = tr.getPeptideRef();
+        // get drift time upper/lower offset (this assumes that all chromatograms
+        // are derived from the same precursor with the same drift time)
+        double drift_target = pep_im_map[pepref];
+        double drift_lower_used = drift_target - im_extraction_win;
```

Comment:
just to double check. is -/+ im_extraction_win correct? I think I recall that somewhere half of a window was used (but might have been mz window).

Reply:
yes, of course. You are right.